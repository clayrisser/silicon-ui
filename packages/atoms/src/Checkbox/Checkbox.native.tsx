import React, { FC, useState, useEffect } from 'react';
import { styled } from 'native-theme-ui';
import {
  CheckBox as NativeBaseCheckbox,
  Item as NativeBaseItem,
  NativeBase
} from 'native-base';
import {
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useItem from '../hooks/useItem';
import {
  StyledCheckBoxProps,
  CheckBoxProps,
  antiForwardCheckBoxPropsKeys,
  splitProps
} from './checkboxProps';

const StyledNativeBaseCheckbox = styled<
  StyledCheckBoxProps,
  NativeBase.CheckBox
>(NativeBaseCheckbox, { forwardPropsBlacklist: antiForwardCheckBoxPropsKeys })(
  compose(border, color, layout, position, shadow, space, typography)
);

const CheckBox: FC<CheckBoxProps> = (props: CheckBoxProps) => {
  const item = useItem();
  const [checked, setChecked] = useState<boolean>(false);

  const { customCheckBoxProps, styledCheckBoxProps } = splitProps(props);

  useEffect(() => {
    if (customCheckBoxProps.checked !== undefined)
      setChecked(customCheckBoxProps.checked);
  }, [customCheckBoxProps.checked]);

  function handleChange(e: any) {
    if (props.onPress) {
      props.onPress(e);
      setChecked(!checked);
    }
  }

  const styledNativeBaseCheckbox = (
    <StyledNativeBaseCheckbox
      {...customCheckBoxProps}
      {...styledCheckBoxProps}
      onPress={handleChange}
      checked={checked}
    />
  );
  if (item.hasItemParent) return styledNativeBaseCheckbox;
  return <NativeBaseItem>{styledNativeBaseCheckbox}</NativeBaseItem>;
};

CheckBox.defaultProps = {
  marginTop: 100,
  checked: false
};

export default CheckBox;
