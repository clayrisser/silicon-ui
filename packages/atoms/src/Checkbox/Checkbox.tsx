import React, { FC, useState } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import {
  CheckBoxProps,
  DetailedHTMLCheckBoxProps,
  splitProps
} from './checkboxProps';

const HTMLCheckBox: StyledComponent<
  DetailedHTMLCheckBoxProps,
  CheckBoxProps,
  object
> = styled.input(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const CheckBox: FC<CheckBoxProps> = (props: CheckBoxProps) => {
  const [checked, setChecked] = useState<boolean>(false);
  const color = useColor(props);
  const {
    customCheckBoxProps,
    styledCheckBoxProps,
    nativeItemProps,
    nativeCheckBoxProps
  } = splitProps({
    ...props,
    color
  });
  function handleChange(e: any) {
    if (props.onPress) {
      props.onPress(e);
      setChecked(!checked);
    }
  }
  return (
    <HTMLCheckBox
      {...styledCheckBoxProps}
      {...nativeItemProps}
      {...nativeCheckBoxProps}
      {...(customCheckBoxProps as any)}
      checked={checked || customCheckBoxProps.checked}
      onChange={handleChange}
      onValueChange={handleChange}
    />
  );
};

CheckBox.defaultProps = {
  // backgroundColor: 'primary',
  // borderRadius: 2,
  // borderWidth: 0,
  // fontFamily: 'body',
  // fontSize: 2,
  // fontWeight: 'body',
  // lineHeight: 'body',
  // paddingBottom: 2,
  // paddingLeft: 2,
  // paddingRight: 2,
  // paddingTop: 2,
  // width: '100%',
  type: 'checkbox',
  checked: false
};

export default CheckBox;
