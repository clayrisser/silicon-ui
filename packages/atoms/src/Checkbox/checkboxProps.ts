import { DetailedHTMLProps, InputHTMLAttributes, ReactNode } from 'react';
import { NativeBase } from 'native-base';
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLCheckBoxProps = DetailedHTMLProps<
  InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
>;

export interface StyledCheckBoxProps
  extends BackgroundProps,
    BorderProps,
    // ColorProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {}

export interface NativeCheckBoxProps extends NativeBase.CheckBox {}

export interface NativeItemProps extends Omit<NativeBase.Item, 'style'> {}

export interface CustomCheckBoxProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  theme?: Theme;
  type?: string;
  children?: ReactNode | string;
  checked?: boolean;
  autoFocus?: boolean;
  indeterminate?: boolean;
}

export interface CheckBoxProps
  extends CustomCheckBoxProps,
    NativeCheckBoxProps,
    NativeItemProps,
    StyledCheckBoxProps {}

export const customCheckBoxPropsKeys = new Set([
  'autoContrast',
  'theme',
  'disabled',
  'checked',
  'onChange'
]);

export const nativeCheckBoxPropsKeys = new Set([
  'error',
  'type',
  'last',
  'onPress',
  'onPressIn',
  'onPressOut',
  'rounded',
  'success',
  'underline',
  'checked',
  'disabled',
  'onValueChange',
  'value'
]);

export const nativeItemPropKeys = new Set<string>([]);

export interface SplitProps {
  customCheckBoxProps: CustomCheckBoxProps;
  nativeCheckBoxProps: NativeCheckBoxProps;
  nativeItemProps: NativeItemProps;
  styledCheckBoxProps: StyledCheckBoxProps;
}

export function splitProps(props: CheckBoxProps): SplitProps {
  const styledCheckBoxProps: { [key: string]: any } = {};
  const customCheckBoxProps: { [key: string]: any } = {};
  const nativeCheckBoxProps: { [key: string]: any } = {};
  const nativeItemProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customCheckBoxPropsKeys.has(key)) {
      customCheckBoxProps[key] = prop;
    } else if (nativeCheckBoxPropsKeys.has(key)) {
      nativeCheckBoxProps[key] = prop;
    } else if (nativeItemPropKeys.has(key)) {
      nativeItemProps[key] = prop;
    } else {
      styledCheckBoxProps[key] = prop;
    }
  });
  return {
    customCheckBoxProps,
    nativeCheckBoxProps,
    nativeItemProps,
    styledCheckBoxProps
  };
}

export const antiForwardCheckBoxPropsKeys = new Set<string>([]);
