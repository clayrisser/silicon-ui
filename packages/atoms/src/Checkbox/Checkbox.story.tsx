import React from 'react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import Checkbox from './Checkbox';
import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import themes from '../themes';
import withThemesProvider from '../../storybook/withThemesProvider';

storiesOf('Checkbox', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemesProvider(themes))
  .addParameters({
    // docs: { page: docs },
    // jest: ['Checkbox.spec.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Checkbox
        disabled={boolean('disabled', false)}
        onPress={action('onPress')}
        checked={boolean('checked', false)}
      />
    </Wrapper>
  ));
