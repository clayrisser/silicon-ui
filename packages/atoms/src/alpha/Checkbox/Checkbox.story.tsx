import React from 'react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import Checkbox from './Checkbox';
import docs from './docs';
import Wrapper from '../../../storybook/Wrapper';
import storiesOf from '../../../storybook/storiesOf';
import withThemeProvider from '../../../storybook/withThemeProvider';

storiesOf('Alpha/Checkbox', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemeProvider)
  .addParameters({
    docs: { page: docs }
    // jest: ['Checkbox.spec.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Checkbox
        disabled={boolean('disabled', false)}
        onPress={action('onPress')}
        checked={boolean('checked', false)}
      />
    </Wrapper>
  ));
