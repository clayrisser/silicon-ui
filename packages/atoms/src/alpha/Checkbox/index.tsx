import Checkbox from './Checkbox';
import { CheckBoxProps } from './checkboxProps';

export { CheckBoxProps };
export default Checkbox;
