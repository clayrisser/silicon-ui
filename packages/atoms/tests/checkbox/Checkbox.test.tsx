import React from 'react';
// import { render, fireEvent } from '@testing-library/react';
// import { render, fireEvent } from '@testing-library/react-native';

import renderer, { act, create } from 'react-test-renderer';
import Checkbox from '../../src/Checkbox';

describe('<Checkbox />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Checkbox />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<Checkbox withProps /> ', () => {
  it('renders with all props correctly', () => {
    const type = 'default';
    const disabled = false;
    const checked = false;
    const tree = renderer
      .create(<Checkbox type={type} disabled={disabled} checked={checked} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<Checkbox with events />', () => {
  it('renders with Event correctly', () => {
    const onPress = jest.fn();

    let component: any;
    act(() => {
      component = create(<Checkbox onPress={onPress} />);
    });
    const rootInstance: any = component.root;
    const checkbox = rootInstance.findAllByType(Checkbox);
    expect(checkbox.length).toBe(1);
    expect(checkbox[0].props.onPress()).toBe(undefined);
  });
});

// describe('<Checkbox onChange />', () => {
//   it('handles onChange', () => {
//     // const handleChange = jest.fn();
//     // const tree = render(<Checkbox onPress={handleChange} />);
//     // fireEvent.press(tree.querySelector('HTMLCheckbox')!);
//     // expect(handleChange).toHaveBeenCalledTimes(1);

//     const handleChange = jest.fn();
//     const { getByPlaceholderText } = render(
//       // MyComponent renders TextInput which has a placeholder 'Enter details'
//       // and with `onChangeText` bound to handleChangeText
//       <Checkbox onPress={handleChange} />
//     );

//     fireEvent(getByPlaceholderText('change'), 'onChangeText', 'ab');
//     expect(handleChange).toHaveBeenCalledWith('ab');
//   });
// });
