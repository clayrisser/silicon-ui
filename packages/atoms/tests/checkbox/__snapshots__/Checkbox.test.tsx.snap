// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders correctly 1`] = `
<input
  checked={false}
  className="css-0"
  onChange={[Function]}
  type="checkbox"
  value={false}
/>
`;

exports[`<Checkbox withProps />  renders with all props correctly 1`] = `
<input
  checked={false}
  className="css-0"
  disabled={false}
  onChange={[Function]}
  type="default"
  value={false}
/>
`;
